ext/parquet/types.go

```diff
@@ -0,0 +1,51 @@
+package parquet
+
+import (
+	"reflect"
+	"strings"
+	"sync"
+)
+
+type structFields map[string]int
+
+var fieldCache sync.Map // map[reflect.Type]structFields
+
+// cachedTypeFields is like typeFields but uses a cache to avoid repeated work.
+//
+// "Inspired" by https://golang.org/src/encoding/json/encode.go
+// Copyright 2010 The Go Authors. All rights reserved.
+func cachedTypeFields(t reflect.Type) structFields {
+	if f, ok := fieldCache.Load(t); ok {
+		return f.(structFields)
+	}
+
+	f, _ := fieldCache.LoadOrStore(t, typeFields(t))
+	return f.(structFields)
+}
+
+func tagName(tag string) string {
+	if pos := strings.Index(tag, ","); pos != -1 {
+		return tag[:pos]
+	}
+	return tag
+}
+
+// "Inspired" by https://golang.org/src/encoding/json/encode.go
+// Copyright 2010 The Go Authors. All rights reserved.
```

Comment:
`func typeFields` seems much bigger in the Go source, does it really need copyright note here?

Reply:
the idea is taken, doesn't matter